Changes to migration guide (#684)
Qiskit/qiskit-ibm-runtime: docs/migrate/migrate-estimator.rst

Calculate expectation values in an algorithm
==============================================

> The Estimator primitive serves a double purpose here:
> 
> * Acts as entry point to quantum systems and simulators by replacing `backend.run()`  (in the case of the `qiskit-ibm-runtime.Estimator`), as well as the tools we had for exact computations (local) that we can now replace with the `qiskit.primitives.Estimator`.
> * Acts as an **algorithmic abstraction**. Recall that to be executable in a real device, the expectation value calculation must represented as a single quantum circuit. This can be done by a series of operations that were previously left to the user. Most users relied on a series of Qiskit Terra tools  (in `qiskit.opflow`) to perform these operations, so this method will be shown in the legacy implementation section.  
> 
>   The new method is much shorter and most of the tools from `qiskit.opflow` used in these examples have been rendered obsolete by the primitives. 

Reviewer: ```suggestion
The role of the `Estimator` primitive is two-fold: on one hand, it acts as an entry point to the quantum devices or simulators, replacing `backend.run()` (or `QuantumInstance.execute()`). On the other hand, it is an **algorithmic abstraction** for expectation value calculations, which removes the need to perform operations to construct the final expectation circuit. This results in a considerable reduction of the code complexity, and a more compact algorithm design. 

The following example uses common tools from the ``qiskit.opflow`` module as a reference for the "legacy way of doing things", but we acknowledge that this is not the only path to perform expectation value calculations. Regardless of what you starting point is, the migration to the `Estimator` primitive will look the same.

```

Reviewer: I am not 100% sold on the wording of the last sentence, hopefully you can find a better-sounding way of transmitting the message.

Problem definition 
-------------------------------

> We want to compute the expectation value of a quantum state (circuit) with respect to a certain operator. Here we are using the H2 molecule and a random circuit as the quantum state:

Reviewer: ```suggestion
We want to compute the expectation value of a quantum state (circuit) with respect to a certain operator. Here we are using the H2 molecule and an arbitrary circuit as the quantum state:
```

> Problem definition 
> ---------------------------
> 
> We want to measure a quantum state:

Reviewer: ```suggestion
```

.. code-block:: python

>     # See https://www.youtube.com/watch?v=1oiYH87vx1A

Reviewer: ```suggestion
```

    from qiskit import QuantumCircuit
    from qiskit.quantum_info import SparsePauliOp

    # Step 1: Define operator
    op = SparsePauliOp.from_list(
        [
            ("II", -1.052373245772859),
            ("IZ", 0.39793742484318045),
            ("ZI", -0.39793742484318045),
            ("ZZ", -0.01128010425623538),
            ("XX", 0.18093119978423156),
        ]
    )

    # Step 2: Define quantum state
    state = QuantumCircuit(2)
    state.x(0)
    state.x(1)

Legacy methods (using opflow)
-----------------------------

> Opflow provided its own classes to represent both operators and quantum states:

Reviewer: Maybe we should add the API reference for opflow here (well, and in general)

    # Convert to expectation value calculation object
    expectation = PauliExpectation().convert(measurable_expression)

    # Note that there are other expectation value methods: MatrixExpectation(), AerPauliExpectation(), 
    # but they are used just like PauliExpectation()

Next, the actual calculation is done by the `CircuitSampler` class, which receives a backend or `QuantumInstance` and the expectation object. Here are some examples of how it can be used:

1. Run locally by using an AerSimulator
*****************************************

.. code-block:: python

   from qiskit.providers.aer import AerSimulator

    # define backend -> local simulator
    simulator = AerSimulator() 

    # inject backend into circuit sampler
    sampler = CircuitSampler(simulator).convert(expectation)

    # evaluate
    expectation_value_2 = sampler.eval().real

    print("sampled: ", expectation_value_2)

2. Run on a remote simulator or real backend
*********************************************

Here we use the `ibmq_qasm_simulator`, but the workflow is the same when using a real device.

.. code-block:: python

    from qiskit import IBMQ

    IBMQ.load_account()

>     provider = IBMQ.get_provider(hub='ibm-q-internal') # here the hub should not be internal

Reviewer: ```suggestion
    provider = IBMQ.get_provider(hub='ibm-q')
```

    backend = provider.get_backend("ibmq_qasm_simulator")

    # inject backend into circuit sampler
    sampler = CircuitSampler(backend).convert(expectation) 

    # evaluate
    expectation_value_4 = sampler.eval().real

    print("sampled: ", expectation_value_4)

> Legacy methods (not using opflow)
> ---------------------------------
> 
> Construct the circuit entirely by yourself and call `backend.run()`.

Reviewer: Replaced by the explanation in the introduction.

```suggestion
```

New method: Use primitives
-----------------------------

Opflow provided its own classes to represent both operators and quantum states:

.. code-block:: python

    from qiskit.opflow import CircuitStateFn, PauliSumOp

    opflow_op = PauliSumOp(op)

>     opflow_state = CircuitStateFn(state) # convert to a state

Author: @ElePT - is the "New method - Use primitives" text correct?  It's the same as "Legacy methods (using opflow).

Reviewer: you're right, this must have got carried over from the previous section in some copy-paste, it's not used anywhere. Let's remove it.

> Opflow provided its own classes to represent both operators and quantum states:
> 
> .. code-block:: python
> 
>     from qiskit.opflow import CircuitStateFn, PauliSumOp
> 
>     opflow_op = PauliSumOp(op)
>     opflow_state = CircuitStateFn(state) # convert to a state

Reviewer: ```suggestion
```

Option 1: Calculate the expectation value exactly
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This can be done with the Estimator primitive in `qiskit.primitives`:

.. code-block:: python

    from qiskit.primitives import Estimator

    estimator = Estimator()

>     result = estimator.run([state], [op]).result().values

Reviewer: This syntax can also be simplified in the following way:

```suggestion
    result = estimator.run(state, op).result().values
```

    print(result)

Option 2: Construct the expectation circuit and sample on a system or simulator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

0. Run locally by using the terra primitive
*********************************************

For the terra primitive, if no shots are specified, it performs an exact calculation. If shots are specified, it performs a shot-based simulation (not quite qasm, as you can see). There is no real legacy alternative for this:

.. code-block:: python

   from qiskit.primitives import Estimator

    estimator = Estimator(options={"shots": 1024})

>     result = estimator.run([state], [op]).result().values

Reviewer: ```suggestion
    result = estimator.run(state, op).result().values
```

    print(result)

1. Run locally by using an AerSimulator 
*********************************************

.. code-block:: python

    from qiskit_aer.primitives import Estimator

    estimator = Estimator(run_options={"shots": 1024})

>     result = estimator.run([state], [op]).result().values

Reviewer: ```suggestion
    result = estimator.run(state, op).result().values
```

    print(result)

2. Run on a remote simulator or real backend
*********************************************

.. code-block:: python
    
    from qiskit_ibm_runtime import QiskitRuntimeService, Estimator

    service = QiskitRuntimeService(channel="ibm_quantum")
    backend = service.backend("ibmq_qasm_simulator")

    estimator = Estimator(session=backend)

>     result = estimator.run([state], [op]).result().values

Reviewer: ```suggestion
    result = estimator.run(state, op).result().values
```

    expectation = PauliExpectation().convert(measurable_expression)

    # Note that there are other expectation value methods: MatrixExpectation(), AerPauliExpectation(), 
    # but they are used just like PauliExpectation()

Next, the actual calculation is done by the `CircuitSampler` class, which receives a backend or `QuantumInstance` and the expectation object. Here are some examples of how it can be used:

**New method - Run locally by using the terra primitive**

0. Run locally by using the terra primitive
*********************************************

For the terra primitive, if no shots are specified, it performs an exact calculation. If shots are specified, it performs a shot-based simulation (not quite qasm, as you can see). There is no real legacy alternative for this:

.. code-block:: python

   from qiskit.primitives import Estimator

    estimator = Estimator(options={"shots": 1024})

    result = estimator.run([state], [op]).result().values
    print(result)

.. _a-legacy-run-aer:

1. Run locally by using an AerSimulator
*****************************************

**Legacy method**

.. code-block:: python

   from qiskit.providers.aer import AerSimulator

>     # define backend -> local simulator
>     simulator = AerSimulator() 
> 
>     # inject backend into circuit sampler
>     sampler = CircuitSampler(simulator).convert(expectation)
> 
>     # evaluate
>     expectation_value_2 = sampler.eval().real
> 
>     print("sampled: ", expectation_value_2)
> 
> **New method**
> 
> .. code-block:: python
> 
>     from qiskit_aer.primitives import Estimator
> 
>     estimator = Estimator(run_options={"shots": 1024})
> 
>     result = estimator.run([state], [op]).result().values
>     print(result)
> 
> .. _a-legacy-run-remote:
> 
> 2. Run on a remote simulator or real backend
> *********************************************
> 
> **Legacy method**
> 
> Here we use the `ibmq_qasm_simulator`, but the workflow is the same when using a real device.
> 
> .. code-block:: python
> 
>     from qiskit import IBMQ
> 
>     IBMQ.load_account()
>     provider = IBMQ.get_provider(hub='ibm-q')
>     backend = provider.get_backend("ibmq_qasm_simulator")
> 
>     # inject backend into circuit sampler
>     sampler = CircuitSampler(backend).convert(expectation) 
> 
>     # evaluate
>     expectation_value_4 = sampler.eval().real
> 
>     print("sampled: ", expectation_value_4)
> 
> **New method**
> 
> .. code-block:: python
>     
>     from qiskit_ibm_runtime import QiskitRuntimeService, Estimator
> 
>     service = QiskitRuntimeService(channel="ibm_quantum")
>     backend = service.backend("ibmq_qasm_simulator")
> 
>     estimator = Estimator(session=backend)
> 
>     result = estimator.run([state], [op]).result().values
>     print(result)

Reviewer: Hello again! I am sorry for the eternal iteration on these examples, but now that I have more experience with migration guides, I see that all of these nested options and methods make the explanation unnecessarily confusing. 

I have a new proposal for restructuring this file, let me know what you think:

```suggestion
Calculate expectation values in an algorithm
==============================================

The role of the ``Estimator`` primitive is two-fold: on one hand, it acts as an entry point to the quantum devices or
simulators, replacing ``backend.run()``.

On the other hand, it is an **algorithmic abstraction** for expectation value calculations, which removes the need
to perform operations to construct the final expectation circuit. This results in a considerable reduction of the code
complexity, and a more compact algorithm design.

.. note::

    The following example uses common tools from the ``qiskit.opflow`` module as a reference for the "legacy way of doing
    things", but we acknowledge that some of you might have used custom code for this task. In that case, you can decide
    between keeping your custom code and replacing ``backend.run()`` with a ``Sampler``, or replacing your custom code with
    the ``Estimator`` primitive.

Problem definition 
--------------------

We want to compute the expectation value of a quantum state (circuit) with respect to a certain operator.
Here we are using the H2 molecule and an arbitrary circuit as the quantum state:

.. code-block:: python

    from qiskit import QuantumCircuit
    from qiskit.quantum_info import SparsePauliOp

    # Step 1: Define operator
    op = SparsePauliOp.from_list(
        [
            ("II", -1.052373245772859),
            ("IZ", 0.39793742484318045),
            ("ZI", -0.39793742484318045),
            ("ZZ", -0.01128010425623538),
            ("XX", 0.18093119978423156),
        ]
    )

    # Step 2: Define quantum state
    state = QuantumCircuit(2)
    state.x(0)
    state.x(1)

.. _a-legacy-opflow:

[Legacy] Convert problem to ``opflow``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

`Opflow <https://qiskit.org/documentation/apidoc/opflow.html>`__ provided its own classes to represent both
operators and quantum states, so the problem defined above would be wrapped as:

.. code-block:: python

    from qiskit.opflow import CircuitStateFn, PauliSumOp

    opflow_op = PauliSumOp(op)
    opflow_state = CircuitStateFn(state)

This step is no longer necessary using the primitives.

.. note::

    For more information on migrating from ``opflow``, see the `opflow migration guide <qisk.it/opflow_migration>`_ .

Calculate expectation values on real device or cloud simulator
---------------------------------------------------------------

[Legacy] Using ``opflow`` + ``backend.run()``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can see the number of steps that legacy workflow involved to be able to compute an expectation
value:

.. note::

    You can replace ``ibmq_qasm_simulator`` with your device name to see the
    complete workflow for a real device.

.. code-block:: python

    from qiskit.opflow import StateFn, PauliExpectation, CircuitSampler
    from qiskit_ibm_provider import IBMProvider

    # Define the state to sample
    measurable_expression = StateFn(opflow_op, is_measurement=True).compose(opflow_state)

    # Convert to expectation value calculation object
    expectation = PauliExpectation().convert(measurable_expression)

    # Define provider and backend (formerly imported from IBMQ)
    provider = IBMProvider()
    backend = provider.get_backend("ibmq_qasm_simulator")

    # Inject backend into circuit sampler
    sampler = CircuitSampler(backend).convert(expectation)

    # Evaluate
    expectation_value = sampler.eval().real

    print("expectation: ", expectation_value)

[New] Using Runtime ``Estimator``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Now, you can notice how the ``Estimator`` simplifies the user-side syntax, which makes it a more
convenient tool for algorithm design.

.. note::

    You can replace ``ibmq_qasm_simulator`` with your device name to see the
    complete workflow for a real device.

.. code-block:: python

    from qiskit_ibm_runtime import QiskitRuntimeService, Estimator

    service = QiskitRuntimeService(channel="ibm_quantum")
    backend = service.backend("ibmq_qasm_simulator")

    estimator = Estimator(session=backend)

    expectation_value = estimator.run(state, op).result().values

    print("expectation: ", expectation_value)

Other execution alternatives (non-Runtime)
------------------------------------------

In some cases, you might want to test your algorithm using local simulation. For this means, we
will should you two more migration paths using non-runtime primitives.

[Legacy] Using Qiskit Aer's Statevector simulator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    from qiskit.opflow import StateFn, PauliExpectation, CircuitSampler
    from qiskit_aer import AerSimulator

    # Define the state to sample
    measurable_expression = StateFn(opflow_op, is_measurement=True).compose(opflow_state)

    # Convert to expectation value calculation object
    expectation = PauliExpectation().convert(measurable_expression)

    # Define statevector simulator
    simulator = AerSimulator(mothod="statevector", shots=100)

    # Inject backend into circuit sampler
    sampler = CircuitSampler(simulator).convert(expectation)

    # Evaluate
    expectation_value = sampler.eval().real

    print("expectation: ", expectation_value)

[New] Using Reference ``Estimator`` or Aer ``Estimator``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Reference ``Estimator`` allows to perform either an exact or a shot-based noisy simulation based
on the ``Statevector`` class in the ``qiskit.quantuminfo`` module.

.. code-block:: python

    from qiskit.primitives import Estimator

    estimator = Estimator()

    result = estimator.run(state, op).result().values

    # for shot-based simulation:
    expectation_value = estimator.run(state, op, shots=100).result().values

    print("expectation: ", expectation_value)

You can still access the Aer Simulator through its dedicated
``Estimator``. This can come in handy for performing simulations with noise models. For more
information on using the Aer Primitives, you can check out this
`VQE tutorial <https://qiskit.org/documentation/tutorials/algorithms/03_vqe_simulation_with_noise.html>`_ .

.. code-block:: python

    from qiskit_aer.primitives import Estimator # all that changes is the import!!!

    estimator = Estimator()

    result = estimator.run(state, op).result().values

    # for shot-based simulation:
    expectation_value = estimator.run(state, op, shots=100).result().values

    print("expectation: ", expectation_value)

```

2. Run on a remote simulator or real backend
*********************************************

Compare this section to :ref:`new-run-remote`.

Here we use the `ibmq_qasm_simulator`, but the workflow is the same when using a real device.

.. code-block:: python

    from qiskit import IBMQ

    IBMQ.load_account()
    provider = IBMQ.get_provider(hub='ibm-q')
    backend = provider.get_backend("ibmq_qasm_simulator")

    # inject backend into circuit sampler
    sampler = CircuitSampler(backend).convert(expectation) 

    # evaluate
    expectation_value_4 = sampler.eval().real

    print("sampled: ", expectation_value_4)

.. new-primitives:

New method: Use primitives
-----------------------------

Compare this section to :ref:`legacy-opflow`.

Opflow provided its own classes to represent both operators and quantum states:

.. code-block:: python

    from qiskit.opflow import CircuitStateFn, PauliSumOp

    opflow_op = PauliSumOp(op)
    opflow_state = CircuitStateFn(state) # convert to a state

.. _new-exact:

Option 1: Calculate the expectation value exactly
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Compare this section to :ref:`legacy-exact`.

> This can be done with the Estimator primitive in `qiskit.primitives`:

Reviewer: ```suggestion
The reference ``Estimator`` in ``qiskit.primitives`` uses the ``qiskit.quantum_info.Statevector`` class to compute expectation values. If no shots are specified, the outcome corresponds to the exact statevector calculation.
```

.. code-block:: python

    from qiskit.primitives import Estimator

    estimator = Estimator()

    result = estimator.run([state], [op]).result().values
    print(result)

.. _new-construct:

Option 2: Construct the expectation circuit and sample on a system or simulator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Compare this section to :ref:`legacy-construct`.

.. _new-construct-run-terra:

0. Run locally by using the terra primitive
*********************************************

> For the terra primitive, if no shots are specified, it performs an exact calculation. If shots are specified, it performs a shot-based simulation (not quite qasm, as you can see). There is no real legacy alternative for this:

Reviewer: ```suggestion
The reference ``Estimator`` in ``qiskit.primitives`` uses the ``qiskit.quantum_info.Statevector`` class to compute expectation values. If the number of shots is specificed, the result will no longer be exact.
```

.. code-block:: python

   from qiskit.primitives import Estimator

>     estimator = Estimator(options={"shots": 1024})
> 
>     result = estimator.run([state], [op]).result().values

Reviewer: ```suggestion
    estimator = Estimator()

    result = estimator.run(state, op, shots=1024).result().values
```

    print(result)

.. _new-run-aer:

1. Run locally by using an AerSimulator 
*********************************************

Compare this section to :ref:`legacy-run-aer`.

.. code-block:: python

    from qiskit_aer.primitives import Estimator

>     estimator = Estimator(run_options={"shots": 1024})
> 
>     result = estimator.run([state], [op]).result().values

Reviewer: ```suggestion
    estimator = Estimator()

    result = estimator.run(state, op, shots=1024).result().values
```
